installer/block_device.go

```diff
@@ -195,3 +252,155 @@ func (bd *BlockDevice) SectorSize() (int, error) {
 
 	return BlockDeviceGetSectorSizeOf(out)
 }
+
+type BlockFrameWriter struct {
+	buf       *bytes.Buffer
+	frameSize int
+	w         io.WriteCloser
+}
+
+// Write buffers the writes into a buffer of size 'frameSize'. Then, when this
+// buffer is full, it writes 'frameSize' bytes to the underlying writer.
+func (bw *BlockFrameWriter) Write(b []byte) (n int, err error) {
+	// TODO -- Add nil guards
+
+	log.Infof("blockFrameWriter: len(b []byte): %d. b: %s", len(b), string(b[:2]))
+
+	// Fill the frame buffer first
+	n, err = bw.buf.Write(b)
+	if err != nil {
+		log.Errorf("bd.buf.Write: %s", err.Error())
+		return n, err
+	}
+
+	if bw.buf.Len() < bw.frameSize {
+		log.Infof("Fewer than chunksize bytes written")
+		return n, nil // Chunk buffer not full
+	}
+
+	log.Info("BlockFrameWriter: Flushing buffer")
+	totWritten := 0
+	nFrames := bw.buf.Len() / bw.frameSize
+	for i := 0; i < nFrames; i++ {
+		n, err = bw.w.Write(bw.buf.Next(bw.frameSize))
+		totWritten += n
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Report the cached bytes as written
+	totWritten += bw.buf.Len() % bw.frameSize
+
+	// Write a full frame, but report only the last byte chunk as written
+	return len(b), nil
+}
+
+// Close flushes the remaining cached bytes, if any.
+func (bw *BlockFrameWriter) Close() error {
+	_, err := bw.w.Write(bw.buf.Bytes())
+	if cerr := bw.w.Close(); cerr != nil {
+		return cerr
+	}
+	return err
+}
+
+// OptimizedBlockDeviceWriter wraps an underlying blockDevice write, however,
+// with the optimization that it compares the bytes passed in to the Write
+// method, with the next len([]byte) bytes (considered a frame) on the block
+// device, and if they match, the write is discarded. The lingo is that only
+// dirty frames are written. Clean ones are discarded.
+type OptimizedBlockDeviceWriter struct {
+	blockDevice BlockDevicer
+	totalFrames int
+	dirtyFrames int
+}
+
+// Write only write 'dirty' frames.
+// Note: a frame-size is always the size 'len(b)'
+func (bd *OptimizedBlockDeviceWriter) Write(b []byte) (n int, err error) {
+
+	log.Infof("Optimized block device writer: len(b): %d, b: %s", len(b), string(b[:2]))
+	log.Infof("Framesize: %d", len(b))
+
+	frameSize := int64(len(b))
+	payloadBuf := make([]byte, frameSize)
+
+	//
+	// Read len(b) bytes from the block-device
+	//
+	n, err = io.ReadFull(bd.blockDevice, payloadBuf)
+	if err != nil {
+		log.Errorf("Failed to read a full frame of size: %d from the block-device", err)
+		return 0, err
+	}
```

Comment:
You should also check what `n` returned. It's not entirely clear what should happen if `n != frameSize` though. It's an unlikely event, but also has very bad consequences if it does. My proposal would be to treat this also as an error.

Reply:
This is already handled in `readFull`. Which will return an error if that is the case.

Reply:
:+1: